lib/bolt/transport/local/shell.rb

```diff
@@ -1,26 +1,179 @@
+              sudo_flags += ["-E"] if options[:environment]
+              sudo_str = Shellwords.shelljoin(sudo_flags)
+              command_str = "#{sudo_str} #{command_str}"
+            else
+              run_as_str = Shellwords.shelljoin(@target.options['run-as-command'] + [run_as])
+              command_str = "#{run_as_str} #{command_str}"
+            end
+          end
+
+          command_arr = options[:environment].nil? ? [command_str] : [options[:environment], command_str]
+
+          # Prepare the variables!
           result_output = Bolt::Node::Output.new
-          result_output.stdout << stdout unless stdout.nil?
-          result_output.stderr << stderr unless stderr.nil?
-          result_output.exit_code = rc.exitstatus
+          in_buffer = options[:stdin] || ''
+          # Chunks of this size will be read/written in one iteration
+          chunk_size = 4096
+          index = 0
+
+          begin
+            inp, out, err, t = Open3.popen3(*command_arr)
+            read_streams = { out => String.new,
+                             err => String.new }
+            write_stream = in_buffer.empty? ? [] : [inp]
+
+            until Process.waitpid(t.pid, Process::WNOHANG)
+              timeout = 0.1
+              # See if we can read from out or err, or write to in
+              ready_read, ready_write, = select(read_streams.keys, write_stream, nil, timeout)
+
+              # Read from out and err
+              ready_read&.each do |stream|
+                begin
+                  if use_sudo
+                    buffer = stream.readpartial(chunk_size)
+                    # Split on newlines, including the newline
+                    lines = buffer.split(/(?<=[\n])/)
+                    # handle_sudo will return the line if it is not a sudo prompt or error
+                    lines.map! { |line| handle_sudo(inp, line) }
+                    read_streams[stream] << lines.join("")
+                  else
+                    read_streams[stream] << stream.readpartial(chunk_size)
+                  end
+                rescue EOFError
+                  strings = { out => 'stdout', err => 'stderr' }
+                  @logger.debug "Read EOF on #{strings[stream]}"
+                end
+              end
+
+              if ready_write && index < in_buffer.length
+                to_print = in_buffer[index..index + chunk_size - 1]
+                # Perform a non-blocking write
+                written = inp.write_nonblock to_print
+                index += written
+              end
+
+              if index >= in_buffer.length && !write_stream.empty?
+                inp.close
+                write_stream = []
+              end
+            end
+          rescue Errno::ECHILD
+            # Returned from waitpid if there is no child process
+            @logger.debug "No child process to wait for; Continuing"
+          end
+          result_output.stdout << read_streams[out]
```

Comment:
It looks like the loop that makes sure the streams are empty after the process exits got removed.

Reply:
I didn't see any issues with data not getting read, including for non-1024 multiple sized strings. I can add it back to be safe though